doc/designs/quic-design/quic-fault-injector.md

```diff
@@ -0,0 +1,507 @@
+we need to do this during MVP in order to be able to observe protocol elements
+sent from the client without modifying them (i.e. in order to confirm that the
+client is behaving as we expect). This will be added if required as we develop
+the tests.
+
+It is expected that the Fault Injector API will expand over time as new
+listeners and helper functions are added to support specific test scenarios. The
+initial API will provide a basic set of listeners and helper functions in order
+to provide the basis for future work.
+
+The following outlines an illustrative set of functions that will initially be
+provided. A number of `TODO(QUIC)` comments are inserted to explain how we
+might exand the API over time:
+
+```` C
+/* Type to represent the Fault Injector */
+typedef struct ossl_quic_fault OSSL_QUIC_FAULT;
+
+/*
+ * Structure representing a parse EncryptedExtension message. Listeners can
+ * make changes to the contents of structure objects as required and the fault
+ * injector will reconstruct the message to be sent on
+ */
+typedef struct ossl_qf_encrypted_extensions {
+    /* EncryptedExtension messages just have an extensions block */
+    unsigned char *extensions;
+    size_t extensionslen;
+} OSSL_QF_ENCRYPTED_EXTENSIONS;
+
+/*
+ * Given an SSL_CTX for the client and filenames for the server certificate and
+ * keyfile, create a server and client instances as well as a fault injector
+ * instance
+ */
+int qtest_create_quic_objects(SSL_CTX *clientctx, char *certfile, char *keyfile,
+                              QUIC_TSERVER **qtserv, SSL **cssl,
+                              OSSL_QUIC_FAULT **fault);
+
+/*
+ * Free up a Fault Injector instance
+ */
+void ossl_quic_fault_free(OSSL_QUIC_FAULT *fault);
+
+/*
+ * Run the TLS handshake to create a QUIC connection between the client and
+ * server.
+ */
+int qtest_create_quic_connection(QUIC_TSERVER *qtserv, SSL *clientssl);
+
+/*
+ * Confirm the server has received a protocol error
+ */
+int qtest_check_server_protocol_err(QUIC_TSERVER *qtserv);
+
+/*
+ * Enable tests to listen for pre-encryption QUIC packets being sent
+ */
+typedef int (*ossl_quic_fault_on_packet_plain_cb)(OSSL_QUIC_FAULT *fault,
+                                                  QUIC_PKT_HDR *hdr,
+                                                  unsigned char *buf,
+                                                  size_t len,
+                                                  void *cbarg);
+
+int ossl_quic_fault_set_packet_plain_listener(OSSL_QUIC_FAULT *fault,
+                                              ossl_quic_fault_on_packet_plain_cb pplaincb,
+                                              void *pplaincbarg);
+
+/*
+ * Helper function to be called from a packet_plain_listener callback if it
+ * wants to resize the packet (either to add new data to it, or to truncate it)
+ */
+int ossl_quic_fault_resize_plain_packet(OSSL_QUIC_FAULT *fault, size_t newlen);
```

Comment:
How does this work if the address of `buf` passed to the callback changes?

Since the size of a datagram is bounded, maybe just have a statically sized buffer and this just changes the logical size? That would be fine. Should be mentioned though.